### ISSUE TITLE
Build Persistence Proposal (SQLite vs. Redis)

### ISSUE THREAD

**Issue by nickzuber** [box/ClusterRunner#401]
Overview
--

This is a proposed design for implementing a storage layer for ClusterRunner so we can persist build history.

### The Problem

As of right now, all of the builds that are created are stored directly on the master service. This causes a few annoyances:

 - We lose all the builds if the master service goes down or we need to restart
 - We are limited to only having a single master service per session

Whenever we initialize a master service, we remove all of the old builds before we start.

```python
if os.path.exists(self._master_results_path):
    fs.async_delete(self._master_results_path)
fs.create_dir(self._master_results_path)
```

We do this because since all the builds are stored on the master, we have no way of getting this information again if the service goes down at any point. We also restart the build ids at 1 so this information would get overwritten anyways. If we move this information off of the master and into some kind of data store, then we wouldn't have to sacrifice this data we have already on disk, and we would still have this information available if we restart the master.

### Possible Solutions

There are a bunch of different ways we can persist this information, but here are a few options that I propose we can go forward with:

 - SQLite Database
 - Redis Data Store

For reference, I ran some [benchmarks to test their performance](https://github.com/box/ClusterRunner/issues/400) in realistic (I hope) situations that they would be used in with ClusterRunner. It's important to note that these benchmarks are not necessarily representative of SQLite and Redis performance in an objective sense. These are two different services, so naturally they would behave differently in the same situations. Also in these particular benchmarks, there are ways to optimize some operations -- however that wouldn't be relevant to the particular implementation we'd be using in ClusterRunner so we don't do that.

SQLite
--

[SQLite](https://www.sqlite.org/about.html) is an easy to use, easy to configure, and lightweight database.

### General Information

SQLite is really easy to set up because the entire database is really just a file. This makes it portable and simple to back up. There's no separate server process involved since the database is in a file -- all operations are just reads and writes to that file. A good quote from the SQLite website is:

> Think of SQLite not as a replacement for Oracle but as a replacement for fopen()

#### Configuration

Since SQLite is essentially just a file, there's no real overhead with set up, configuration, or managing. The only requirement for using SQLite is to be able to read from disk. No additional set up or overhead required in ClusterRunner is nice to have.

#### Efficient Queries with Autoincrement

By default, SQLite [indexes items with a field called `ROWID`](https://sqlite.org/autoinc.html). This field is autogenerated starting at 1 and continuing until the largest possible integer (9223372036854775807). This field autoincrements and is _incredibly_ performant to use in queries. For reference, access queries for builds are [about 4x faster using `ROWID` in SQLite than Redis](https://github.com/box/ClusterRunner/issues/400#rowid). This behavior is already what we want for generating build `id`s so we can get this right out of the box.

The reason queries with `ROWID` are so fast is because, as [the SQLite documentation states](http://sqlite.org/lang_createtable.html#rowid):

> The data for rowid tables is stored as a B-Tree structure containing one entry for each table row, using the rowid value as the key. **This means that retrieving or sorting records by rowid is fast**. Searching for a record with a specific rowid, or for all records with rowids within a specified range is around twice as fast as a similar search made by specifying any other PRIMARY KEY or indexed value.

#### True Concurrency

SQLite is not great with concurrency. Any amount of services can _read_ from the database at any given time, but only one can _write_ at any given time. The writes are queued and performed synchronously. This could come into play if we have multiple master services trying to write builds to the database at the same time, however we should note that we can expect a writer only spends a few milliseconds writing to disk, so this probably won't be an issue in practice for our use case.

#### Maybe Race Conditions with `ROWID`

SQLite only allows one write to the database at a time which means that we hopefully don't have to worry about race conditions. The one issue that I thought of was if we ever had two masters trying to create a build at the same time. This wouldn't be an issue with inserting the build, however there _could_ be some trouble when it comes to getting that build's `id` immediately after it's inserted.

There's a feature that allows us to call `lastrowid` from the SQLite cursor to get the last `id` that was inserted into the table. This is used to get the build's `id` right after we insert it since the `id` comes from the autoincremented `ROWID` within our SQLite table. If more than one master inserted a build before we called `lastrowid`, we could end up in a situation where we think two different builds have the same `id`. 

However, this can be avoided if each master has its own connection to the database. This is because [a SQLite connection cursor is the one who knows the `ROWID` of the last inserted row that _it_ inserted](https://stackoverflow.com/a/6242813/5055063). I think (and correct me if I'm wrong) that one master instance will not be trying to create multiple builds at once. If this is true, then if each master instance try to create a build at the same time, each of their SQLite connections will know the last `ROWID` that _they themselves_ inserted, therefore **not** creating a race condition on getting the correct build `id`.

If for some reason this ever becomes a problem later down the line, we could always make the BuildStore class a singleton and place all incoming build requests into a queue that sequentially handles insertions one at a time. I highly doubt it'd ever come to this though, but if it does we could solve the problem.

### How It Would Look

Whenever we start up ClusterRunner, we'd want to connect to our database in a BuildStore and connect an instance of that BuildStore to our master. Also as a side note, a cool thing about SQLite limit select queries is that they use an `offset, limit` pattern just like we do in our pagination, so no extra logic needed!

#### build_store.py
```python
import sqlite3
from typing import List

from app.master.build import Build
from app.util.exceptions import ItemNotFoundError
from app.util.log import get_logger


class BuildStore(object):
    """
    Build storage service that stores and handles all builds.
    """

    def __init__(self):
        self._logger = get_logger(__name__)
        self._sqlite_db = sqlite3.connect('clusterrunner.db')
        self._sqlite_connection = self._sqlite_db.cursor()

        # Check if builds table exists, if not then create it
        if not self._sqlite_connection.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='builds'").fetchone()
            sqlite_connection.execute('CREATE TABLE builds(build TEXT)')

    def get(self, build_id: int) -> Build:
        """
        Returns a build by id
        :param build_id: The id for the build whose status we are getting
        """
        try:
            self._sqlite_connection.execute('SELECT rowid, * FROM builds WHERE rowid=?', (build_id,))
            build_id, build = sqlite_connection.fetchone()
        except:
            raise ItemNotFoundError('Invalid build id: {}.'.format(build_id))

        return build

    def get_range(offset: int=None, limit: int=None) -> List['Build']:
        """
        Returns a list of all builds.
        :param offset: The starting index of the requested build
        :param limit: The number of builds requested
        """
        try:
            if offset and limit:
                self._sqlite_connection.execute('SELECT rowid, * FROM builds limit ?, ?', (offset, limit))
            else:
                self._sqlite_connection.execute('SELECT rowid, * FROM builds')

            results = self._sqlite_connection.fetchall()
        except:
            # Raise selection failure

        return [build for build_id, build in results]

    def add(self):
        """
        Add new build to collection
        """
        # Create build here
        # Builds will have a build_id of `None` when first created and this build_id will be set after insertion
        build = # ...
        try:
            self._sqlite_connection.execute('INSERT INTO builds VALUES(?)', (build,))
            self._sqlite_db.commit()
            build_id = self._sqlite_connection.lastrowid  # Gets the build_id of build we just inserted
            # Finish build creation by assigning its build_id
            # ...
        except:
            # Raise insertion failure

```

#### master_service.py
```python
class ClusterMaster(ClusterService):
    """
    The ClusterRunner Master service: This is the main application class that the web framework/REST API sits on top of.
    """

    API_VERSION = 'v1'

    def __init__(self):
        self._logger = get_logger(__name__)
        self._master_results_path = Configuration['results_directory']
        self._all_slaves_by_url = {}
        self._all_builds_by_id = OrderedDict()
        self._scheduler_pool = BuildSchedulerPool()
        self._build_request_handler = BuildRequestHandler(self._scheduler_pool)
        self._build_request_handler.start()
        self._slave_allocator = SlaveAllocator(self._scheduler_pool)
        self._slave_allocator.start()
        # Assign instance of BuildStore to master service
        self._build_store = BuildStore()

        SlavesCollector.register_slaves_metrics_collector(lambda: self.all_slaves_by_id().values())

```

Redis
--

[Redis](https://redis.io/) is a performant in memory data store that's primarily used for caching information.

### General Information

Since Redis is an in memory data store, this means all of its data is kept in memory. We can write this information to disk, and Redis will do this automatically, after certain intervals of time. Since data is kept in memory, this makes it fast when it comes to setting and getting key values. This also means that the builds are organized in a different way than they would be in SQLite -- in Redis, a key is assigned a value so we'd have a build `id` assigned to a build whereas in SQLite we have a row in a table that has a build `id` and its respective build. Subtle, but still a difference. 

#### Expire Keys

Redis has an expire feature that can expire data after a certain amount of time. This can be pretty useful for us since we'd have to clear out old builds eventually -- with the expire feature, we can help automate some of this process. The one downside to this feature, however, is that [any attempts to manipulate the contents of a build will result in the expire tag to render itself void](https://redis.io/commands/expire). This means we have to be very careful with how we handle the data stored or else we risk removing the expire tag which can be annoying.

#### Performance

Redis is really performant and good at doing insertion and access operations. While looking at [some benchmarks](https://github.com/box/ClusterRunner/issues/400), it's clear to see the power behind Redis and how quickly it can do things. This does, however, come at a cost regarding persistence, which I talk more about later, but in general Redis does a quicker job of storing and looking up information compared to SQLite _depending on what we use for indices_.

#### Configuration Overhead

Redis needs its own server process running for it to work, so this requires installation and set up overhead. Also if we wanted to configure things such as disk persistence timing intervals we'd have to make sure that's set up correctly as well. We'd also need to keep tabs on and maintain the Redis server to make sure it stays up and doesn't have any issues.

#### Deterministic Persistence

Redis actually [isn't great at deterministically persisting data](https://redis.io/topics/persistence). Since it's an in memory data store, all of the builds would naturally be kept in memory. There _is_ a persistence feature where the data store is written to disk, but this isn't super reliable. That's because the intervals which we can write to disk are limiting in the sense of performance. The more frequently we write to disk, the less performant our Redis cache will be. And if we decide to write to disk less often, we run into the risk of losing data stored in memory if our Redis server goes down at any point for any reason.

I'd argue that we can't afford to lose this kind of data. Let's assume user A attempts to create a build. ClusterRunner then creates that build and let's say it's the 100th build in our data store. Therefore, it gets the build `id` of 100. User A now expect the build with an `id` of 100 to have that specific build. Now, let's say something happens and our Redis server goes down before this build was written to disk. Later, user B comes along and creates a build. Now _this_ build is given an `id` of 100 since we lost all the history of the original 100th build. When user A comes to check on her build 100, she get's a totally different build. I think this kind of technical behavior is major and something we can't afford to have happen in our system.

So we have the option of writing to disk more frequently, but even if we shorten that window, there's _still_ an opportunity for this data loss happening (it just makes it less likely). The only way to prevent this is to write to disk every time a new build is created, which could be very expensive. A quote from the Redis website:

> RDB needs to fork() often in order to persist on disk using a child process. Fork() can be time consuming if the dataset is big, and may result in Redis to stop serving clients for some millisecond or even for one second if the dataset is very big and the CPU performance not great.

The main reason we'd want to use Redis over SQLite is for its performance, and doing something like this would make that advantage almost moot.

### How It Would Look

For installing Redis, you can [refer to this guide](https://redis.io/topics/quickstart).

First we would need to spin up a Redis server.

```bash
❯ redis-server /path/to/redis.conf
```

Then we connect to our Redis service in our BuildStore and master services can just interface with that BuildStore to create and get builds.

#### build_store.py
```python
import redis
from typing import List

from app.master.build import Build
from app.util.exceptions import ItemNotFoundError
from app.util.log import get_logger


class BuildStore(object):
    """
    Build storage service that stores and handles all builds.
    """

    def __init__(self):
        self._logger = get_logger(__name__)
        # Connect to Redis service
        self._redis_connection = redis.Redis('localhost', db=0)

    def get(self, build_id: int) -> Build:
        """
        Returns a build by id
        :param build_id: The id for the build whose status we are getting
        """
        build = self._redis_connection.get(build_id)
        if build is None:
            raise ItemNotFoundError('Invalid build id: {}.'.format(build_id))

        return build

    def add(self):
        """
        Add new build to collection
        """
        # Create build here
        build = # ...
        self._redis_connection.set(build_id, build)

```

We would also need a way to increment the `build_id` counter correctly and start it off at the last `build_id` each time we restart. This could be something like a lookup for the largest key value in our Redis store when ClusterRunner starts up and starting our counter from there.

This is something we don't really seem to get out of the box and would have to handle ourselves, which isn't a big deal but means more work.

### Other Things to Think About

Regardless of which approach we take, there are some universal things we need to solve and think about.

#### Serialization

A build is stored as a dictionary of information that looks something like this:

```python
{
    'num_subjobs': 0,
    'status': '',
    'details': 'null',
    'failed_atoms': 'null',
    'result': 'null',
    'num_atoms': 'null',
    'id': 1,
    'request_params': {
        'build_project_directory': '',
        'url': '',
        'job_name': '',
        'type': ''
    },
    'state_timestamps': {
        'queued': 1,
        'preparing': 1,
        'building': 'null',
        'error': 'null',
        'finished': 'null',
        'prepared': 'null',
        'canceled': 'null'
    },
    'error_message': 'null'
}
```

Both SQLite and Redis can't store nested dictionaries like this out of the box. One option we have is to serialize the entire dictionary as a string using `json.dumps` and then storing that string, and when we want to get the build back we can use `json.loads` to deserialize. This is kind of expensive and we should avoid doing it if a better exists.

One other idea is to decide what information from this dictionary is important and effectively reduce the amount of things we want to store. If there is any unneeded information here that we don't really care about in the context of persisting builds, leave a comment and we can try to reduce this data structure to something simpler.

### Conclusion and Final Thoughts

With all things considered, I don't think Redis is the right tool for the job here. If we use SQLite `ROWID`s for build `id`s, lookup queries can be as fast, if not faster than Redis lookup times. Writing builds is also very similar in terms of speed when it comes to Redis and SQLite. So when it comes to performance, I don't think there's much of a difference -- if anything, we might prefer SQLite.

Also, I think the point that Redis can't guarentee _all_ information will be written to disk is a huge issue with how we use ClusterRunner. If we're going to persist builds, we need to be able to persist _all_ the builds.

Especially with the set up overhead, SQLite seems like a more practical and easy to use solution. Less maintenance and configuration required with SQLite than Redis helps keep ClusterRunner as a tool that's easy to work with, easy to set up, and most importantly easy to maintain.

At the end of the day, Redis is great for temporary storage but we need a more permanent solution. SQLite seems like a very performant and practical tool for our use case, and I think we should go ahead with that approach.

**Comment by wjdhollow** [comment]
[Why not MySQL?](http://mcfunley.com/choose-boring-technology)

> SQLite is really easy to set up because the entire database is really just a file. This makes it portable and simple to back up. There's no separate server process involved since the database is in a file -- all operations are just reads and writes to that file. 

MySQL is really easy to setup too. Plus, it is managed by DBA's so you get things like replication for free. Implementation wise, it would be the same as SQLite (assuming we use SQLAlchemy) with the exception of what driver we use. 

As a service, you should assume that the host availability SLA's are pretty low because, more than likely, it can be in a public/private cloud. That means you are susceptible to disk failures, hypervisor issues or a network issue with connecting to your provider. There is no guarantee that the files on disk will be there unless you are actively backing it up. So the file as persistence approach would require another component. Unless of course, you don't care about backing up the file; in which case, kind of defeats the purpose of persisting at all :neckbeard: 

> Redis has an expire feature that can expire data after a certain amount of time. This can be pretty useful for us since we'd have to clear out old builds eventually

I wouldn't list this as a good thing because clearing out builds is something we would have to do to run an in-memory database. With MySQL, we could store a lot of rows. 

4 bytes for primary key
4 bytes for num subjob
4 bytes for num atoms
3 bytes per date time * 6 = 18
255 bytes per string * 5 = 1275

So that would be 1305 bytes per row. For a million rows that would be 1.3 Gb. Which is nothing in terms of storage size.

> Redis is really performant and good at doing insertion and access operations.

We are not using ClusterRunner at a scale where a user would notice the difference in read/write speeds. I think MySQL is a stronger choice here because as you mentioned, indexing is key. Pun intended. 

```
Both SQLite and Redis can't store nested dictionaries like this out of the box. One option we have is to serialize the entire dictionary as a string using json.dumps and then storing that string, and when we want to get the build back we can use json.loads to deserialize. This is kind of expensive and we should avoid doing it if a better exists.
```

What's the matter? You don't like schemas 😈 

I'd assume we would use a schema if we went with MySQL. A downside of doing the dictionary dump approach is that you are hand tying yourself to only querying by key. The ability to query is huge and a big plus for MySQL. By going with a json dump approach, you are effectively removing the possibility of easily querying your DB. Creating the schema is pretty straightforward. The ClusterRunner data model has been consistent for a few years now so I don't see that as a barrier. 

Could you add a section to this proposal that considers a MySQL (+ schema) as an option? For me, I think SQLite is off the table, but I would be open to Redis, considering that all the tradeoffs were considered. Though, one middle ground w/ SQLite would be to abstract the driver in SQLAlchemy and expose it as a configuration option. That way it makes it easy to adopt and easy to productionize.

**Comment by ethomas2** [comment]
I agree with pretty much everything you said. I think sqlite/mysql are better suited for the job. The risk of losing user data far outweighs the benefit of the performance of redis. For me, redis is out.

I also agree with will that mysql is better than sqlite. Using mysql would allow us to run 2 cluster runner masters both pointed at the same db. It also gives us replication for free, we don't have to worry about a machine dying and losing the sqlite file on that machine. As long as we use sqlalchemy it's easy to support both. My vote is to default to using sqlite so that setup stays as simple as (this)[http://www.clusterrunner.com/docs/quickstart/] (no need to set up a mysql server), but add a config option that lets you switch to mysql. 

Out of curiosity, did you consider mongodb at all? I think I might prefer mysql/sqlalchemy over mongodb because
1. The ease of setup/configuration/deployment
2. I don't want to have to set up and maintain mongodb servers, but we already have a mysql server running
3. I'm not familiar with mongodb, and I don't want to slow down your code reviews by having to research mongo db stuff 

but ... I think this is the exact situation that mongodb is made for. Namely, I want to store and query json blobs. You solve the serialization problem, you get replication, I _believe_ you get auto incrementing, you get querying. All the good stuff.

Don't worry about actually writing up something to support or oppose mongodb. Just if you had already thought about it and chosen to discard it, I'd like to hear your thoughts.

**Comment by vartikaz** [comment]
@nickzuber if we have two CR masters, how would that affect the choice between SQLite, Redis, MySQL.

**Comment by wjdhollow** [comment]
@vartikaz You bring up good points. I think we should define an SLA around build persistence. Those requirements greatly impact the choice. If persistence is not an issue, then replication becomes a nice-to-have which then makes Redis a more viable option. Partly, why I mentioned I'd be more open to that than SQLite. 

@nickzuber I'm not that familiar with Redis. Can you include in your proposal what sort of convention we would adopt for keys?  In your example, you used "$id", but isn't that limiting? We also have data models for slaves. I think it would be more sense if we used the "$model:$id:$field" convention. So maybe something like build:13:subjob:12:atoms:0. The reason I don't think that the data blob approach will work is because subjobs are independent events and their status is currently updated by separate threads. How would Redis handle two threads trying to update the same blob? To my understanding, that is one of it's weaknesses, though I could be wrong. If it is a weakness, we should list it as such.

@ethomas2 MongoDB sounds cool. I'd like to see that comparison.

**Comment by ethomas2** [comment]
> Also, ease of deployment and keeping setup easy are great points. However, I would prioritize performance over those. WDYT

I prioritize performance very low. I don't really care if I submit a new build to cluster runner and it takes a full second before I get a response. I'd care if latency in the master was response for the UI lagging taking  ~.5 seconds or more to load, but I probably wouldn't prioritize fixing that.

Even if performance was important, we experience absurdly low load on cluster runner master. I'm sure mysql can handle it. Here's a graph of requests per second in cluster runner over the last week. It maxed out at 45 req/s and even that was unusually high. Usually it's something like 10 req/s.
<img width="1269" alt="screen shot 2017-08-08 at 4 15 15 pm" src="https://user-images.githubusercontent.com/13339709/29098621-3676e8aa-7c55-11e7-935c-5b1236abf259.png">

**Comment by ethomas2** [comment]
If we had to worry about our aws bill then I'd care about using extra CPU cycles, which is _a_ performance concern. But I still probably wouldn't care about latency. At least not until we get way more load in cluster runner.

**Comment by tjlee0909** [comment]
Great write up!

As @ethomas2 said, the read and write performances at that level are NOT significant for ClusterRunner. We had a similar debate at the very beginning of ClusterRunner when we were deciding whether to write it in Python or Go or C++, and we came to the conclusion that we should choose whatever is the easiest for us to implement and maintain ClusterRunner in. ClusterRunner is meant to reduce test feedback from something like 30 hours down to 15 minutes; shaving off a few hundred milliseconds from a 15 minute build isn't very meaningful. And I also somehow doubt that whatever extra CPU cycles we end up burning through won't be especially meaningful either (that said, I don't know how exactly drastic this CPU usage may be for different options).

Ease of use (including deployment and system administration) is a huge selling point of ClusterRunner (not that we are popular enough for this to matter...yet). As simple as the service is right now, we already struggle with the operational usability of ClusterRunner. We keep on saying that we'll work on revamping the deployment stuff for ClusterRunner, and when we finally get to it I'd really rather not have to worry about another service that we have to worry about keeping up and properly starting up/shutting down on demand.

I personally don't want to, or find the need, of having a full-blown MySQL service running in our initial version of this project. Whatever cost we have to maintain and know about another running service doesn't merit whatever benefit it brings based on anything I've seen yet in our encounters with ClusterRunner.

I'm personally still for going with the Sqlite database (entirely for it's ease of maintainability and deployment) for our initial, and possibly long-standing, iteration of this project. If it ever turns out that we change our mind in the future, then we should be able to re-use almost all of our usage of SqlAlchemy to point to a full-fledged replicated MySQL cluster.

The data on ClusterRunner is not important enough, in my opinion, to trade off any more additional complexity or even engineering time over. On average, we restart ClusterRunner (and consequently wipe all build data) once every two-four weeks--and we have done this for the last three years . While not a good experience for our customers, it is not the dev-equivalent of a business-ending event (e.g.: losing git refs from github enterprise or gitosis). Also, regarding implementing disk backups of Sqlite files -- we could absolutely do that if we wanted, but I don't even suspect that it's worth it. If this really becomes a serious enough problem that the simple solution won't do, then lets just swap the Sqlite DB with a MySQL cluster.

And regarding the JSON dump vs schema discussion -- definitely should go with what @wjdhollow suggested and create a MySQL schema, for all of the benefits that Will listed.

**Comment by tjlee0909** [comment]
Oh, and it'd be very easy to have the default behavior of it to use Sqllite, but have optional configuration settings for hostname, port, and credentials for a MySQL cluster in clusterrunner.conf.

Thank you SQLAlchemy!

**Comment by nickzuber** [comment]
I think SQLite is a good first step for this problem and will give us all the behavior we want in a performant way assuming that we only have a single master service. We won't have to worry about multiple master services until later down the line, but is something we should consider in our design.

@tjlee0909 brings up a good point with MySQL and SQLAlchemy that when we get to the point of wanting to have multiple master services and we "outgrow" SQLite, we could use MySQL and communicate with that database using SQLAlchemy. Since we'd already be using SQLite, the migration to this regarding the CR codebase would be very small, and we'd be able to reuse most of the same logic. 

Because of these reasons, I still think we should use SQLite and not necessarily worry about multiple masters until we reach that point. In the meantime, using SQLite solves the problems that we're setting out to solve here, _and_ also gives us the ability to scale to using MySQL with SQLAlchemy when we get to the point of wanting to use multiple master services with relatively little extra work in the CR codebase.

**Comment by wjdhollow** [comment]
I'm on board with SQLAlchemy and SQLite. The ORM provides us enough of an abstraction layer that we could easily switch to MySQL. SQLAlchemy even has an in-memory DB option http://docs.sqlalchemy.org/en/latest/core/engines.html#sqlite, which is nice. 

My vote is for SQLAlchemy. I'm fine with hashing out the schema in code review, a separate issue, or over multiple commits. Since it's all on file/ memory, we don't have to do any prod migrations, so I'm comfortable with moving super fast here.

**Comment by nickzuber** [comment]
Recap
--

Problems to solve
--

One of the main things we want to solve is being able to restart CR. This means if we shut down in the middle of a build, we'd want to be able to continue that build when we start up CR again.

To do this, it'd be nice to have an easy way to know which builds (if any) are not completed when CR starts up.


Approaches
--

There are really two approaches here:

 1. Store builds in a database and cache them locally when we want to use them
 2. Keep all the builds in memory using a data store like Redis


SQL Databse w/ SQLAlchemy
--

With this design, we'd still have builds stored in memory in the `BuildStore` class. These builds that are in memory are either previous incompleted builds that we picked back up on restart, new builds being requested in the current session, or older builds that are requested at any point during the current session.

Using an SQL database like SQLite, we need to think about _when_ we want to write builds to the database and when we get them. Since our main priority here is to recover builds after a restart and continue working on any uncompleted builds, we can figure out some essential times we'd need to write and read from the database:

 - **On start up** — query database for any uncompleted builds, deserialize them into `Build` objects, store them in memory in the `BuildStore` class, have `cluster_master` check for unfinished builds in the `BuildStore` class, add any it finds to the build request queue.
 - **On shut down** — serialize all builds in the `BuildStore` class, query database to update all respective builds.
 - **On new build request** — create `Build` object like normal but with no `build_id`, serialize `Build` object, store serialized build into database, get its `build_id` that was just generated and assign it back to the `Build` object, store the build in memory on the `BuildStore` class.
 - **On get build request** — check in memory cache in the `BuildStore` class for the requested build, if it can't be found query the database for that build, derserialize it in a `Build` object, add the build to the `BuildStore` class in memory cache, return that build.

### Serialization & Tricky Bits

The tricky bit about this approach is actually serializing and deserializing a `Build` object without losing any of its data and state. However, having to break down a `Build` object and serialize it is unavoidable, even with a Redis approach.

There are some shortcuts we can make where we some data to initialize other parts of a `Build` when we want to deserialize it. For example, instead of storing all of the data for a `project_type`, we just call `self.generate_project_type()` when we initialize our `Build` object.

#### Subjobs

Another tricky part about maintaining state is the subjob status. In a `Build` object, we have `_unstarted_subjobs` and `_finished_subjobs` to organize which subjobs we need to work on in this build. Since subjobs themselves don't have any idea if they're complete or not, we could keep another column in a Subjobs table to mark them as complete or incomplete. When deserializing a build, if a subjob is marked as complete, it'll be sorted into the `_unstarted_subjobs` queue and the `_finished_subjobs` if it's marked as complete.

#### Preparing State

One thing to think about is when a build is in the preparing state. Any given build can only be prepared once; this process fetches the project and sets up some files, initializes the subjob queues, and creates the results directories. If CR is shutdown in the middle of this process, what's the best way to handle this? Since the build never finished getting prepared, I think it makes sense to treat it like an unprepared build we store it in the database. 

This comes into play when we initialize a `Build` object on a CR start up, we want to make sure we initialize the build with the correct `SingleUseCoin()` value to mark if it should be prepared or not.

We could describe the status of a build like:

```python
class BuildSchema(Base):
    """Describes the status of a Build"""
    __tablename__ = 'builds'

    build_id = Column(Integer, primary_key=True, autoincrement=True)
    completed = Column(Boolean)  # Set to True when self._status() is FINISHED
    prepared = Column(Boolean)   # Set to True when self._status() is not PREPARING
```

### Advantages

 - **Simple** — Assuming we can serialize and deserialize a `Build` object without losing state or information, it'd be really simple to maintain. Using the `BuildStore` class, all we'd have to worry about is when to save and add builds. For the builds in memory, if we want we can even use a smarter cache to purge builds we don't want to cache locally anymore and save more memory that way.

### Disadvantages
 - **Potentially expensive queries** — Since there's so much data a build has, there's a change that trying to update a lot of builds at once will be expensive and have a noticable performance hit. However, this can be mitigated by updating builds as they complete and only updating uncompleted builds on a shut down, or some similar approach.

**Comment by ethomas2** [comment]
Looks good. Minor comments:

1. We discussed offline that this design is making the BIG assumption that when cluster runner master is restarted, the slaves are all restarted as well. That's fine for now since the master _requires_ slaves to be restarted anyway (because of the session-id), but that should be explicitly called out in the design.

2. Can you flesh out the schema a lot more? I want to know exactly what tables we'll be using and exactly what columns those tables have.

3. In the bit of the build schema you do have, you separated `completed` and `prepared`. I think there should just be one column named `state` of type `Enum` whose values are all possible build states. This makes illegal states unrepresentable (e.g. completed=True but prepared=False). I'll comment more on this when you have a fleshed out schema.

**Comment by nickzuber** [comment]
@ethomas2 This is what the schema currently looks like (welcome to feedback and improvements). The goal of the schema is to completely describe each part of a Build object so we can reconstruct the same Build object using this info.

Note that this is without your suggestion of having one column for `state` instead of splitting up `prepared` and `completed` -- I like this approach much better especially if we don't want to try and start up uncompleted builds yet.

```python
class BuildStateSchema(Base):
    """Describes the state of a Build"""
    __tablename__ = 'builds'

    build_id = Column(Integer, primary_key=True, autoincrement=True)
    completed = Column(Boolean)
    prepared = Column(Boolean)


class BuildMetaSchema(Base):
    """Describes all the Build attributes that are basic datatypes"""
    __tablename__ = 'build_metas'

    build_id = Column(Integer, primary_key=True)
    artifacts_tar_file = Column(String)
    artifacts_zip_file = Column(String)
    error_message = Column(String)
    postbuild_tasks_are_finished = Column(String)
    setup_failures = Column(Integer)
    timing_file_path = Column(String)
    # call `self.generate_project_type()` after init

class BuildArtifactSchema(Base):
    """Describes the _build_artifact attribute of a Build"""
    __tablename__ = 'build_artifacts'

    build_id = Column(Integer, primary_key=True)
    build_artifact_dir = Column(String)


class FailedArtifactDirectoriesSchema(Base):
    """Describes the list of _failed_artifact_directories (List[str]) from BuildArtifact"""
    __tablename__ = 'failed_artifact_directories'

    uid = Column(Integer, primary_key=True, autoincrement=True)
    build_id = Column(Integer)
    failed_artifact_directory = Column(String)


class FailedSubjobAtomPairsSchema(Base):
    """Describes the list of _failed_subjob_atom_pairs (List[int, int]) from BuildArtifact"""
    __tablename__ = 'failed_subjobs_atom_pairs'

    uid = Column(Integer, primary_key=True, autoincrement=True)
    build_id = Column(Integer)
    subjob_id = Column(Integer)
    atom_id = Column(Integer)


class BuildRequestSchema(Base):
    """Describes the build_request attribute of a Build. Use these values to initialize a BuildRequest object."""
    __tablename__ = 'build_requests'

    build_id = Column(Integer, primary_key=True)
    type = Column(String)
    url = Column(String)
    branch = Column(String)
    job_name = Column(String)
    # other optional attributes...


class BuildFsmSchema(Base):
    """Describes the _state_machine in a Build Object"""
    __tablename__ = 'build_fsms'

    build_id = Column(Integer, primary_key=True)
    state = Column(String)
    queued = Column(Float)
    finished = Column(Float)
    prepared = Column(Float)
    preparing = Column(Float)
    error = Column(Float)
    canceled = Column(Float)
    building = Column(Float)


# project_type = build.generate_project_type()
# job_config = self.project_type.job_config()
# Subjob(build_id, subjob_id, project_type, job_config, atoms)
# ^-- this will set all the atoms to `NOT_STARTED` fyi
class SubjobsSchema(Base):
    """Describes the subjobs associated with a Build"""
    __tablename__ = 'subjobs'

    uid = Column(Integer, primary_key=True, autoincrement=True)
    subjob_id = Column(Integer)
    build_id = Column(Integer)
    completed = Column(Boolean)


class AtomsSchema(Base):
    """Describes the subjobs associated with a Build"""
    __tablename__ = 'atoms'

    uid = Column(Integer, primary_key=True, autoincrement=True)
    atom_id = Column(Integer)
    build_id = Column(Integer)
    subjob_id = Column(Integer)
    command_string = Column(String)
    expected_time = Column(Float)
    actual_time = Column(Float)
    exit_code = Column(Integer)
    state = Column(String)
```

**Comment by josephharrington** [comment]
@nickzuber
Re: the sqlite vs. redis metrics -- very interesting! I'm interested in how the metrics change if we utilize a schema in the sqlite db instead of json serializing/deserializing.

I also noticed in your sample code for sqlite that you only commit once after all the inserts. Perhaps it would be more realistic to commit after each insert, although I'm actually more concerned about the *update* operation performance than the insert, especially when separate request threads are trying to update (how much is it affected by needing to create a separate connection for each new request thread?). I imagine that most of the activity on build will be updates instead of creates or reads. For example, a single build would be created once, read into memory once if the master is restarted, but its subjobs will be updated potentially thousands of times (depending on how many atoms it has) as each atom finishes. 

One thing in general I'd like more clarity on is the multithreading support in sqlite. I'm assuming that because of the current design of clusterrunner (single-threaded web server) that we won't run into issues around this -- the writes would naturally be serialized because API calls are serialized. Are there any concerns here if we ever switched to a multithreaded web server?

I had a few comments about Redis, but they've mostly been addressed by other comments above. The general opinion from everyone seems to be that it is not the right tool for this use case, which I agree with.

More inline feedback:

> "We are limited to only having a single master service per session"

- Build persistence alone is not enough to enable a multi-master architecture, especially if we're using sqlite. 


> "SQLite is not great with concurrency. ... This could come into play if we have multiple master services trying to write builds to the database at the same time,"

- I don't think it's realistic to expect running mutiliple masters if we're using sqlite. How would two master services on different hosts share the same file? If we're running multiple masters I think it is a given we'd need a separate persistence service like mysql or redis.


> "A build is stored as a dictionary of information that looks something like this:..."

 - What about the atom data? (Exposed via subjob and atom API endpoints) This is used heavily by the console output dashboard which is the main view that users have into their build status.
 - For nested dicts in sqlite, I imagine those could be separate tables (as Will and TJ also suggested). (It looks like your most recent comment with a proposed schema includes this idea.) 

We should clarify right now if we want this design to address HA concerns / support multiple masters. I think it should not. You've mentioned it in a few places, but if that is actually a design goal then there's much more to fill in. For example, your point about potentially storing a subset of build data would not be an option if we were supporting multiple masters.

My strong opinion is that the main goal of this design should be simply to allow us to restart or ugprade the master and slave services with the *option* of not losing any build data. Things that implies:
 - We would not support restarting the master without restarting slaves
 - We would not support HA for the master service
 - We do not make guarantees about how long we'll persist builds for (for example we might want to wipe out the db if there's a breaking change to schema etc.)

I don't think we should go overboard with build persistence by making guarantees that we'll persist builds indefinitely. If that's an incidental byproduct of the design then that's fine, but I still feel strongly that long-term build result persistence should be handled by a separate service (Jenkins, etc.).

Even though the current behavior is not great, users deal pretty well with losing build data in the current world. It's never ideal to involuntarily lose data, but I suspect in 99% of cases that build results are only ever needed for a few hours after the build runs. 


@wjdhollow
> "For me, I think SQLite is off the table"

- Do you still think that after the rest of the conversation here? This decision obviously depends on the requirements. If the requirement is to only allow us to restart/update the master without losing data then I don't think sqlite is out.

@ethomas2
> "Using mysql would allow us to run 2 cluster runner masters both pointed at the same db."

- This might be true but I think this is more complicated than it sounds. You'd also need to share slaves between those masters which I haven't seen a proposal for and is not at all trivial (the master services need to share a single idle slaves queue, or that whole piece needs to be redesigned). Until we have a more complete proposal for HA I don't think we should include that in our design considerations.

> "My vote is to default to using sqlite so that setup stays as simple as this (no need to set up a mysql server), but add a config option that lets you switch to mysql."

 - +1, although supporting both makes testing considerably more difficult. Sqlalchemy makes the interface consistent, but I don't think the behaviors are completely interchangeable, especially when it comes to threading. I agree with TJ's suggestion to only go with sqlite for now. I'm not completely sure, but I believe the sqlite requirements are stricter than mysql, so if we solve all issues with sqlite then switching to mysql later might be easy (but I don't think we should start with mysql if we also want to support sqlite).

@vartikas
> "Also, ease of deployment and keeping setup easy are great points. However, I would prioritize performance over those. WDYT?"

 - I would prioritize performance only as far as it affects test execution time. As TJ mentioned, ease of setup was a *major* design goal of clusterrunner and I think remains one of its strengths. We can probably now satisfy our ease-of-setup requirement by focusing on docker-based deployments -- that allows us to essentially use as many separate services as we want without complicating quickstart setup (we could use something like docker-compose), but we'd need to allocate time for docker stuff. If we can avoid it I'd prefer not to compromise ease-of-setup in the short term.


@ethomas2
> "I prioritize performance very low. I don't really care if I submit a new build to cluster runner and it takes a full second before I get a response."

 - I agree with that. Performance should really only be considered as far as it affects test execution time, and I don't think *any* of the proposals so far would have a significant affect on that. As long as we are not doing something silly like adding a 1 second operation for each atom I think all of the solutions proposed here are fine performance-wise.

@nickzuber
> "One of the main things we want to solve is being able to restart CR. This means if we shut down in the middle of a build, we'd want to be able to continue that build when we start up CR again."

 - I think being able to resume builds is a *huge* additional requirement that we should not try to address right now. That would require persisting much more state (unstarted/started subjobs, assigned slaves... basically everything in the BuildScheduler class) and also require changes to the slaves to be able to be robust against master restarts. We'd need to address the currently unresolved issue of version differences between the master and slaves since it would be more likely the master could be upgraded ahead of the slaves. We'd need to persist all slave state for the master in addition to build state, including the idle slave queue. We'd also need a way to gracefully stop subjob assignment/slave allocation, etc during a shutdown to avoid race conditions around what we've persisted. Overall, I think supporting resuming in-progress builds is much more work and would add much more complexity than the original requirement of only persisting finished builds. I'm not sure the effort/complexity will ever be worth the value of being able to restart while builds are running. It would be much easier to have some sort of graceful shutdown mode that queues up builds and waits for running builds to finish before restarting.

> "Since our main priority here is to recover builds after a restart and continue working on any uncompleted builds, we can figure out some essential times we'd need to write and read from the database: ..."

 - What about as atoms finish? As I mentioned above I think this is the most frequent state change.

> "However, having to break down a Build object and serialize it is unavoidable"

 - I'm not sure what you mean here. If we use sqlalchemy (as seems to be the most popular opinion) the Build object becomes a db model object that can be saved/loaded without our app code needing to worry about serialization.

---

To summarize my input:
 - I support going with sqlite via sqlalchemy.
 - We should not include any HA/multi-master considerations in this design. I believe the requirements for that feature go beyond merely what we choose as a data store.
 - I think the goal for this design should only be to enable restarting the master+slaves without losing *completed* build data (finished + canceled + errored builds). If we see any builds in the db when the master starts up, those should be marked as errored.
 - The current expectation is that slaves are restarted when the master is restarted. I don't see a reason to change that for this design.

**Comment by nickzuber** [comment]
@josephharrington 

I believed we talked a bit offline yesterday about this but I'm gonna answer some of your questions here:

> I'm actually more concerned about the update operation performance than the insert, especially when separate request threads are trying to update

Yeah this is why we want to be smart about _when_ we update a build in the database and why it'd be worthwhile to have a local cache of builds that are being worked on. I propose we only write to the database in a few different scenarios:

 - When a build is first created (this is how we get its `build_id` too)
  - The build will be marked as "in progress" so if its never finished because of a shutdown mid build or something), we know not to pretend that it's completed if someone wants to look at it. This also allows us to potentially pick up uncompleted builds and finish them when the master restarts, but we won't be actually doing that in this proposal.
 - When a build completes
  - Ideally we'd want to update the build in our database after the postbuild tasks are finished. This will also allow us to purge completed builds from our local cache too, since we won't be doing much with them once they've completed.
 - When cluster master service is shut down gracefully.
  - This allows us to save any builds in progress in their current state to potentially picked back up later when the master service starts up. Again, we won't be implementing this part in this proposal, but it will give us the ability to do this in the future if we want to.

Alternatively, we could keep another dictionary of all the SQLAlchemy build models and update those as a build updates to mitigate the amount of queries we execute, but I don't think this is necessary and adds more complexity by essentially having two copies of a build (one being the actualy Build object and the other being the SQLAlchemy model).


> Are there any concerns here if we ever switched to a multithreaded web server?

It seems like so long as we make sure [each thread has its own database connection](https://sqlite.org/threadsafe.html), we should be fine with using a multithreaded web server w/ SQLite.

> I think being able to resume builds is a huge additional requirement that we should not try to address right now.

Yeah I spoke more with Evan and TJ offline about this and the issues with the slave services and everything else lead me to believe we shouldn't be trying to restart any uncompleted builds in this proposal. I still think there's benefit to saving partially completed builds though, even if we're not going to try and restart them just yet.

> What about as atoms finish? As I mentioned above I think this is the most frequent state change.

Yeah in my original post I was making the false assumption that we'd want to be storing the simple JSON API representation instead of an entire build. So https://github.com/box/ClusterRunner/issues/401#issuecomment-322887577 I lay out what I have so far for the schema for storing a complete build with all its attributes, subjobs, and atoms so we'd be able to recreate the build in its exact state from where we saved it.

> I'm not sure what you mean here. If we use sqlalchemy (as seems to be the most popular opinion) the Build object becomes a db model object that can be saved/loaded without our app code needing to worry about serialization.

Sorry I realize this was kind of vague. What I meant was we can't avoid serializing and deserializing a Build object.
